db/compaction/compaction_job.h

```diff
@@ -225,6 +227,26 @@ class CompactionJob {
   // consecutive groups such that each group has a similar size.
   void GenSubcompactionBoundaries();
 
+  // Get the number of planned subcompactions against various limit. For
+  // round-robin, it may call ReserveSubcompactionResources to reserve threads
+  // and update extra_num_subcompaction_threads_reserved_ if necessary
+  uint64_t InitializeNumPlannedSubcompactions(uint64_t max_output_files_limit,
+                                              uint64_t ranges_size_limit);
+
+  // The number of extra reserved threads will be stored in
+  // extra_num_subcompaction_threads_reserved__
+  void ReserveSubcompactionResources(int extra_required_subcompactions);
+
+  // Additional threads may be reserved during ReserveSubcompactionResources()
+  // if num_actual_subcompactions is less than num_planned_subcompactions.
+  // Additional threads will be released and the bg_compaction_scheduled_ or
+  // bg_bottom_compaction_scheduled_ will be updated if they are used.
+  // DB Mutex lock is required.
+  void ShrinkSubcompactionResources(uint64_t num_unnecessary_subcompactions);
+
+  // Release all reserved threads and update the compaction limits.
+  void ReleaseSubcompactionResources();
```

Comment:
Sorry, I'm a bit confused by the expression here. Now we will have 5 functions?:
* `void InitializeSubcompactionResources(uint64_t, uint64_t)`
* `uint64_t GetSubcompactionLimit()`: called by `InitializeSubcompactionResources`
* `IncreaseSubcompactionResources()` : renaming `ReserveSubcompactionResources()
* `ShrinkSubcompactionResources()`: take away the condition around `ShrinkSubcompactionResources` by taking the min as the input
* `ReleaseSubcompactionResources()`

What is the use of the first one? Should not it can return the number of planned compactions?

> The former can call the latter at the start (when `extra_num_subcompaction_threads_reserved_ == 0`) and use that to decide how much resources need to be reserved. 

Also confused about this. `extra_num_subcompaction_threads_reserved_`  is determined after calling the limit and together with permitted `max_subcompaction` we can compute how much resources need to be reserved.

Reply:
> Sorry, I'm a bit confused by the expression here. Now we will have 5 functions?:

Yes.

> * `void InitializeSubcompactionResources(uint64_t, uint64_t)`

Right, or `void AcquireSubcompactionResources(uint64_t, uint64_t)` per other comment.

> * `uint64_t GetSubcompactionLimit()`: called by `InitializeSubcompactionResources`, what does this function return?

It returns `std::max(std::uint64_t(1), static_cast<uint64_t>(c->max_subcompactions()) +  extra_num_subcompaction_threads_reserved_ `. This value has the following meaning at different times:

- At the start of `InitializeSubcompactionResources()`, it can be called to get `max_subcompactions_limit`
- After `InitializeSubcompactionResources()`, it can be called to get `num_planned_subcompactions`

> * `IncreaseSubcompactionResources()` : renaming `ReserveSubcompactionResources()
> * `ShrinkSubcompactionResources()`: take away the condition around `ShrinkSubcompactionResources` by taking the min as the input. For your information, `ReleaseThreads` also takes the mutex lock and releases it even if the input is zero. Here I may also add an early return if the input number is 0.
> * `ReleaseSubcompactionResources()`
>
> What is the use of the first one? Should not it return the number of planned compactions? I guess here we just need to rename `InitializeNumPlannedSubcompactions`, am I right?

Right about rename. I'd also make it return `void` and query the `num_planned_subcompactions` from `GetSubcompactionLimit()`.

> > The former can call the latter at the start (when `extra_num_subcompaction_threads_reserved_ == 0`) and use that to decide how much resources need to be reserved. 
>
> Also confused about this. `extra_num_subcompaction_threads_reserved_`  is determined after calling the limit and together with permitted `max_subcompaction` we can compute how much resources need to be reserved. 

Calling `GetSubcompactionLimit()` at that point is just another way to calculate `max_subcompactions_limit`. I realize it's a trivial amount logic duplication saved so this isn't important to me.

> > The latter can be greater than the former in case `max_subcompactions` already permitted more subcompactions to run than needed.
>
> I don't understand why permitted `max_subcompactions`can play a role here. The input of shrink is larger than the reserve, which is because the number of actual compactions is not as large as the number of planned one.

Hm. `max_subcompactions` plays a clearer role in shrink's input being larger than reserve's **return value**. It feels counterintuitive that we try to shrink more resources than we ever reserved. This happens when `num_planned_subcompactions - num_actual_subcompactions > extra_num_subcompaction_threads_reserved_`. I thought that'd happen when `num_actual_subcompactions < max_subcompactions`.

Reply:
Thanks a lot for your detailed clarification!